composer/core/state.py

```diff
@@ -540,14 +547,117 @@ def state_dict(self) -> Dict[str, Any]:
 
         return state_dict
 
-    def load_model_state(self, state_dict: Dict[str, Any], strict: bool):
+    def _apply_required_algorithms(
+        self,
+        state: Dict[str, Any],
+        logger: Logger,
+        exclude_algorithms: Optional[List[str]] = None,
+        algorithm_passes: Optional[List[AlgorithmPass]] = None,
+    ):
+        """Applies required algorithms which haven't been specified and aren't in the exclude list.
+
+        Args:
+            state (Dict[str, Any]): State from checkpoint.
+            logger (Logger): Logger to use.
+            exclude_algorithms (List[str], optional): List of algorithm names to exclude. (default: ``None``)
+            algorithm_passes (List[AlgorithmPass], optional): A list of algorithm passes to apply to autoloaded algorithms
+                to sort them into the correct order. (default: ``None``)
+        """
+        import composer.algorithms as algorithms  # type: ignore imports used in `eval(representation)`
+
+        # Get repr of existing algorithms
+        state_algos = {}
+        for algo in self.algorithms:
+            if algo.required_on_load():
+                if type(algo) not in state_algos:
+                    state_algos[type(algo)] = []
+                state_algos[type(algo)].append(algo.__repr__())
+
+        # Gather algorithms to apply
+        missing_algos = set()
+        missing_algo_names = []
+        missing_algo_reprs = []
+        for algo_name, serialized_value in state['algorithms']:
+            if hasattr(algorithms, algo_name) and getattr(algorithms, algo_name).required_on_load():
```

Comment:
any chance you can maybe add a small comment above some of these if blocks to clarify exactly what the case is and what you are doing in that case? I find the triple nested if statements to be a bit hard to follow.